src/Bicep.Core/TypeSystem/DeployTimeConstantVisitor.cs

```diff
@@ -0,0 +1,130 @@
+        // whether a runtimeValue is allowed for the currentSymbol
+        private bool runtimeValueAllowed;
+
+        private DeployTimeConstantVisitor(SemanticModel.SemanticModel model, IDiagnosticWriter diagnosticWriter)
+        {
+            this.model = model;
+            this.diagnosticWriter = diagnosticWriter;
+            this.runtimeValueAllowed = true;
+            this.currentSymbol = "";
+        }
+
+        // entry point for this visitor
+        public static void ValidateDeployTimeConstants(SemanticModel.SemanticModel model, IDiagnosticWriter diagnosticWriter)
+        {
+            new DeployTimeConstantVisitor(model, diagnosticWriter).Visit(model.Root.Syntax);
+        }
+
+        public override void VisitResourceDeclarationSyntax(ResourceDeclarationSyntax syntax)
+        {
+            this.runtimeValueAllowed = false;
+            base.VisitResourceDeclarationSyntax(syntax);
+            this.runtimeValueAllowed = true;
+        }
+
+        public override void VisitModuleDeclarationSyntax(ModuleDeclarationSyntax syntax)
+        {
+            this.runtimeValueAllowed = false;
+            base.VisitModuleDeclarationSyntax(syntax);
+            this.runtimeValueAllowed = true;
+        }
+
+        public override void VisitObjectPropertySyntax(ObjectPropertySyntax syntax)
+        {
+            bool priorRuntimeValueAllowed = this.runtimeValueAllowed;
+            // runtimeValueAllowed should only be false if we are going through a module or resource's objectPropertySyntax
+            if (!this.runtimeValueAllowed && syntax.Key is IdentifierSyntax keyIdentifier) 
+            {
+                this.currentSymbol = keyIdentifier.IdentifierName;
+                // right now we only check for the name property (using resourceNamePropertyName but the name is same for modules)
+                if (!keyIdentifier.IdentifierName.Equals(LanguageConstants.ResourceNamePropertyName, System.StringComparison.Ordinal))
+                {
+                    this.runtimeValueAllowed = true;
+                }
+            }
+            base.VisitObjectPropertySyntax(syntax);
+            // restore the value prior to the visit
+            this.runtimeValueAllowed = priorRuntimeValueAllowed;
+        }
+
+        public override void VisitPropertyAccessSyntax(PropertyAccessSyntax syntax)
+        {
+            if (!this.runtimeValueAllowed)
+            {
+                if (syntax.BaseExpression is VariableAccessSyntax variableAccessSyntax)
+                {
+                    // validate only on resource and module symbols
+                    // TODO these switch cases are so redundant.....
+                    switch (model.GetSymbolInfo(variableAccessSyntax))
+                    {
+                        case ResourceSymbol resourceSymbol:
+                            if ((resourceSymbol.Type is ResourceType resourceType) &&
+                            (resourceType.Body is ObjectType resourceBodyObj))
```

Comment:
>ObjectType [](start = 50, length = 10)

There's a function in `TypeAssignmentVisitor` that could be useful here. It's called `UnwrapType` - could be worth moving it to some common location.

Also, resourceType.Body can be a `DiscriminatedObjectType`. The resource type `Microsoft.Resources/deploymentScripts@2020-10-01` is like that. When you add tests, make sure you add some with that resource type.

Reply:
I'm not able to reproduce the `DiscriminatedObjectType` case